### PR TITLE
Take into account generic types returned by SubTypesSelector

### LINKED ISSUE #1981: Generic base classes
When I have 

    public class Base
    {
        public int BaseProperty { get; set; }
    }

    public class Generic<T> : Base
    {
        public T GenericProperty { get; set; }
    }

    public class Final : Generic<int>
    {
        public int FinalProperty { get; set; }
    }

The generated swagger.json doesn't have a reference to Base.  After enabling 'UseAllOfForInheritance' it still doesn't work? When I takeout either the Base class or Generic in both cases it works fine. But the combined case as presented above doesn't work.

Now when I use NSwag studio to generate a client the Final class no longer derives from the base class.

### PR DESCRIPTION
**Pull request by MikeFH**
Fixes #1981

### CODE REVIEW

**Review by qsdfplkj — APPROVED** [review]
I would like to test this with a prerelease package.

**Comment by JacobAtchley** [comment]
+1 on this. Am experiencing the same issue with Polymorphic Generics

**Comment by idealser** [comment]
Any updates on it? Same issue happens to my projects.

**Comment by dotkarl** [comment]
I'm maintaining a project that faces the same issue. Is there any reason why this PR hasn't been merged yet, even though it was approved in May 2021? Has it already been fixed by some other PR, or is there still some problem with this code?

**Comment by martincostello** [comment]
Thanks for contributing - if you'd like to continue with this pull request, please rebase against the default branch to pick up our new CI.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SchemaGenerator/SchemaGenerator.cs [hunk 1]
```diff
@@ -175,7 +175,9 @@ private bool IsBaseTypeWithKnownTypesDefined(DataContract dataContract, out IEnu
 
             if (dataContract.DataType != DataType.Object) return false;
 
-            var subTypes = _generatorOptions.SubTypesSelector(dataContract.UnderlyingType);
+            var subTypes = _generatorOptions
+                .SubTypesSelector(dataContract.UnderlyingType)
+                .Where(type => !type.IsGenericTypeDefinition); ;
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SchemaGenerator/SchemaGenerator.cs:180]:
```suggestion
                .Where(type => !type.IsGenericTypeDefinition);
```

Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SchemaGenerator/SchemaGenerator.cs:180]: Could you add a comment explaining why generic types are skipped?

I'm new to the internals of the codebase and it isn't immediately clear to me.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SchemaGenerator/SchemaGenerator.cs [hunk 2]
```diff
@@ -414,8 +416,16 @@ private bool IsKnownSubType(DataContract dataContract, out DataContract baseType
...
 
-            if (baseType == null || baseType == typeof(object) || !_generatorOptions.SubTypesSelector(baseType).Contains(dataContract.UnderlyingType))
+            if (baseType == null ||
+                baseType == typeof(object) ||
+                !_generatorOptions.SubTypesSelector(baseType).Any(type =>
+                    type == dataContract.UnderlyingType ||
+                    (type.IsGenericTypeDefinition && dataContract.UnderlyingType.IsConstructedGenericType && type == dataContract.UnderlyingType.GetGenericTypeDefinition())
+                ))
+            {
+
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SchemaGenerator/SchemaGenerator.cs:426]:
```suggestion
```

File: test/Swashbuckle.AspNetCore.Newtonsoft.Test/SchemaGenerator/NewtonsoftSchemaGeneratorTests.cs [hunk 3]
```diff
@@ -161,6 +161,30 @@ public void GenerateSchema_GeneratesReferencedDictionarySchema_IfDictionaryTypeI
...
+        [Fact]
+        public void GenerateSchema_GeneratesBaseType_IfGenericInHierarchy()
+        {
+            var subject = Subject(
+                configureGenerator: c => c.UseAllOfForInheritance = true
+            );
+
+            var schemaRepository = new SchemaRepository();
+
+            var referenceSchema = subject.GenerateSchema(typeof(SubTypeOfGeneric), schemaRepository);
```
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.Newtonsoft.Test/SchemaGenerator/NewtonsoftSchemaGeneratorTests.cs:173]: Can we also assert on the names of the properties that are found? Or is that not worth doing?

**Review by martincostello — COMMENTED** [review]
A few minor comments - also please fix the (new since rebase) xunit analyser warnings.

**Comment by martincostello** [comment]
Would the changes here also fix #2507?

**Pull request closed without merging**